### ISSUE TITLE
Just one cpu core in use, until I use numpy...

### ISSUE THREAD

**Issue by melonista** [pytorch/pytorch#841]
Hi,

I followed the installation instructions for linux that you have on GitHub, including the optional dependencies part.

I run the following code:

```
import torch
a = torch.rand(10000,100000)
b = torch.rand(100000,10000)
c = a.mm(b)
```

However, when I run that script in a Linux machine where I installed python with Anaconda, and I also installed mkl and anaconda accelerate, that script uses just one core. I have tried compiling from source, and also installing pytorch with "conda install", and also not installing the accelerate library, but it never uses more than one core during that script.

However, if I first run this script, which uses all cores:
```import numpy as np
a = np.random.rand(10000,100000)
b = np.random.rand(100000,10000)
c = a.dot(b)
```

And afterwards I run again the torch one, then it uses all cores of the machine.

What am I missing? Do I have to set some flag?

I have also observed that, if I don't run the numpy script first, when I train a neural network, it seems to use all cores at some point when computing the gradients, but not during the forward pass.

This is what the compiler says while compiling:

```
-- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_gf: not found
-- Checking for [mkl_gf - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_gf: not found
-- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_intel_lp64: /home/user/local_installs/anaconda3/lib/libmkl_intel_lp64.so
--   Library mkl_gnu_thread: not found
-- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_intel_lp64: /home/user/local_installs/anaconda3/lib/libmkl_intel_lp64.so
--   Library mkl_intel_thread: /home/user/local_installs/anaconda3/lib/libmkl_intel_thread.so
--   Library mkl_core: /home/user/local_installs/anaconda3/lib/libmkl_core.so
--   Library pthread: /usr/lib64/libpthread.so
--   Library m: /usr/lib64/libm.so
-- Looking for cblas_sgemm
-- Looking for cblas_sgemm - not found
-- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - pthread - m]
--   Library mkl_intel: not found
-- Checking for [mkl_intel - mkl_intel_thread - mkl_core - pthread - m]
--   Library mkl_intel: not found
-- Checking for [mkl_gf_lp64 - mkl_sequential - mkl_core - m]
--   Library mkl_gf_lp64: not found
-- Checking for [mkl_gf - mkl_sequential - mkl_core - m]
--   Library mkl_gf: not found
-- Checking for [mkl_intel_lp64 - mkl_sequential - mkl_core - m]
--   Library mkl_intel_lp64: /home/user/local_installs/anaconda3/lib/libmkl_intel_lp64.so
--   Library mkl_sequential: /home/user/local_installs/anaconda3/lib/libmkl_sequential.so
--   Library mkl_core: /home/user/local_installs/anaconda3/lib/libmkl_core.so
--   Library m: /usr/lib64/libm.so
-- Looking for cblas_sgemm
-- Looking for cblas_sgemm - found
-- MKL library found
-- Performing Test BLAS_F2C_DOUBLE_WORKS
-- Performing Test BLAS_F2C_DOUBLE_WORKS - Failed
-- Performing Test BLAS_F2C_FLOAT_WORKS
-- Performing Test BLAS_F2C_FLOAT_WORKS - Success
-- Found a library with BLAS API (mkl).
-- Found a library with LAPACK API. (mkl)
```

**Comment by melonista** [comment]
It seems I have to manually set mkl cores before using torch. For some reason numpy does that for me.

```
import mkl
mkl.set_num_threads(56)

import torch

a = torch.rand(10000,100000)
b = torch.rand(100000,10000)
c = a.mm(b)
```

**Comment by MatthewKleinsmith** [comment]
Sometimes mkl caps the number of threads below your system's max. See mkl's cap via `mkl.get_max_threads()`. To increase the cap to your actual max, use `export MKL_DYNAMIC=FALSE` in bash before running Python. Find your actual max via `echo $(nproc)` in bash.

I came across this solution in this issue:
https://github.com/ContinuumIO/mkl-service/issues/2

**Comment by yinrong** [comment]
> import mkl
> mkl.set_num_threads(56)

after mkl.set , cpu still can NOT take > 100%
